### PR TITLE
add crypto service

### PR DESCRIPTION
**Pull request by rkondratowicz**
### Change description ###
Just the service with tests. Not used by API yet.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    private static final String CIPHER = "AES/GCM/NoPadding";
+    private static final Length KEY_LENGTH = Length.BITS_128;
+    private static final Length IV_LENGTH = Length.BITS_96;
+    private static final Length GMC_TAG_LENGTH = Length.BITS_128;
+
+    public static byte[] encrypt(String input, String password) {
+
+        IllegalArgument.throwIf(() -> input == null, "Input can't be null");
+        IllegalArgument.throwIf(() -> Strings.isNullOrEmpty(password), "Password can't be empty");
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: Any requirements on the password?

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import static com.google.common.base.Charsets.UTF_8;
+
+public class CryptoService {
+
+    private static final String CIPHER = "AES/GCM/NoPadding";
+    private static final Length KEY_LENGTH = Length.BITS_128;
+    private static final Length IV_LENGTH = Length.BITS_96;
+    private static final Length GMC_TAG_LENGTH = Length.BITS_128;
+
+    public static byte[] encrypt(String input, String password) {
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: Is there a library that could this for us and be less risky for us to implement?

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: There's not much going on here, we just provide configuration for `javax.Crypto` (key length, IV length, AES mode etc).

The only custom things is that IV is concatenated with cipher text.

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java (hunk 1 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: It would be a good idea add some validation of the password (length) to request validation. Not critical, though, as our clients are other HMCTS services (responsible for using strong passwords), not individuals.

Comment (timja-kainos) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: probably a min length would be enough...

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: I'd rather validate that on API level

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                new GCMParameterSpec(GMC_TAG_LENGTH.bits(), generateIv())
+            );
+
+            return Bytes.concat(
+                cipher.getIV(),
+                cipher.doFinal(input.getBytes(UTF_8))
+            );
+
+        } catch (GeneralSecurityException exc) {
+            throw new GeneralSecurityRuntimeException(exc);
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: Worth adding a message, saying we failed to encrypt the data. I imagine that the message in exc may be quite enigmatic.

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java (hunk 1 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: Alternatively we can require a 128bit key to be sent from the client. Instead of deriving it from the password

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        }
+    }
+
+    public static String decrypt(byte[] input, String password) {
+
+        IllegalArgument.throwIf(() -> input == null || input.length == 0, "Input can't be empty");
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: You could consider using Strings.isNullOrEmpty, like you did below. Looks a bit nicer.

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: it's a byte array here

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            byte[] cipherText = Arrays.copyOfRange(input, IV_LENGTH.bytes(), input.length);
+
+            Cipher cipher = Cipher.getInstance(CIPHER);
+            cipher.init(
+                Cipher.DECRYPT_MODE,
+                deriveKey(password),
+                new GCMParameterSpec(GMC_TAG_LENGTH.bits(), iv)
+            );
+
+            return new String(cipher.doFinal(cipherText), UTF_8);
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: Maybe we should rely on the encoding set up for the application? It's probably going to still be UTF_8, but at least there would be one place where we specify it. What do you think?

File: src/test/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+import org.junit.Test;
+
+import java.util.stream.Stream;
+import javax.crypto.AEADBadTagException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+
+public class CryptoServiceTest {
```
Comment (KarolJastrzebski) [inline src/test/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoServiceTest.java]: Missing assertions for:
- plain == decrypted
- plain != encrypted

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+        } catch (GeneralSecurityException exc) {
+            throw new GeneralSecurityRuntimeException(exc);
+        }
+    }
+
+    private static SecretKey deriveKey(String password) throws NoSuchAlgorithmException {
+        byte[] hashedPassword = MessageDigest.getInstance("SHA-512").digest(password.getBytes());
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: I'm wondering if adding some salt here would increase security. If the user has multiple drafts and some drafts are the same, they would also look the same encrypted. On the other hand, does it matter?

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: The cipher text will be different. Random IV is used every time.

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/Length.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package uk.gov.hmcts.reform.draftstore.service.crypto;
+
+public enum Length {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/Length.java]: good idea 👍

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java (hunk 1 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: Those passwords could be named `secrets`. To us they're secret pieces of information, based on which we will encrypt the data.

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java (hunk 7 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: Right, you add it to the message.

File: src/test/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoServiceTest.java (hunk 6 above)
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoServiceTest.java]: added

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java (hunk 4 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoServiceTest.java (hunk 6 above)
Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoServiceTest.java]: We could also add a test for checking if an exception would be thrown if we used an empty password or an empty input (although maybe we should allow for input to be empty) and a test checking if encrypted input of a very short length is long enough in the encrypted form.

File: src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+            return new String(cipher.doFinal(cipherText), UTF_8);
+
+        } catch (GeneralSecurityException exc) {
+            throw new GeneralSecurityRuntimeException(exc);
+        }
+    }
+
+    private static SecretKey deriveKey(String secret) throws NoSuchAlgorithmException {
+        byte[] hashedPassword = MessageDigest.getInstance("SHA-512").digest(secret.getBytes());
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/reform/draftstore/service/crypto/CryptoService.java]: hashedSecret?
